---
title: "Feline Bile Microbiome in Biliary Disease: Data Curation"
author: "BJC"
date: "12/7/2022"
output: html_document
---

## Import and Harmonize Data

Load packages.
```{r load-packages}
library(biomformat); packageVersion("biomformat")
library(dada2); packageVersion("dada2")
library(decontam); packageVersion("decontam")
library(ggplot2); packageVersion("ggplot2")
```

Read in the microbiome profile data.
```{r read-biom}
fn.biom <- "../Data/feature-table.biom" # CHANGE ME IF working directory is different than Rmd location
biom.hash <- read_biom(fn.biom) 
st.hash <- t(as(biom_data(biom.hash), "matrix")) # Now, rows are samples and columns are ASVs
```

This biom file was generated by QIIME2 running the q2-dada2 plugin, and the ASVs are stored as hashes rather than as the sequences themselves. Using the representative sequence fasta to translate the hashes back to DNA sequences.
```{r asv-sequences}
fn.seqs <- "../Data/dna-sequences.fasta" # CHANGE ME IF working directory is different than Rmd location
sq <- getSequences(fn.seqs)
if(!identical(colnames(st.hash), names(sq))) stop("Mismatch between hashed ASV IDs.")
st <- st.hash
colnames(st) <- sq # Sequence table now "DADA2 format": Rows=samples, Columns=ASVs, corresponding row/colnames
```

Read in the sample metadata.

```{r}
fn.meta <- "../Data/bile_map.csv" # CHANGE ME IF working directory is different than Rmd location
df <- read.csv(fn.meta, header=TRUE)
rownames(df) <- df$SampleID
```

Check for consistency between sequence table and metadata table.

```{r}
identical(sort(rownames(df)), sort(rownames(st))) # FALSE
```

```{r}
rownames(df)[!rownames(df) %in% rownames(st)]
rownames(st)[!rownames(st) %in% rownames(df)]
```

Samples "726-518" and "727-520" are in the sequence table, but not in the metadata table.

Coordinating the sequence table and metadata table (and removing samples without metadata).

```{r}
sam <- rownames(df)
df <- df[sam,]
st <- st[sam,]
if(!identical(rownames(df), rownames(st))) stop("Mismatched samples in df and st.")
```

A brief glance at the data we are working with.

```{r}
dim(df)
```

```{r}
dim(st)
```

77 samples. 93 metadata columns. 10,000+ ASVs.

## Curate Data

Inspect metadata columns.

```{r}
colnames(df)
```

Begining with the columns that we will use to try to distinguish between contaminants and real sample sequences:

* `$DNA_Conc`: The DNA concentration prior to library normalization. ND is below limit of detection
* `$Library_Conc`: DNA concentration after library normalization.
* `$SampleType`: Positive, or NegativeSampling/NegativeIsolation/NegativeLibrary, all of which are negative samples of different types

```{r}
table(df$SampleType, useNA="ifany")
```

16 total negative controls, most of which are (13/16) sampling negatives (blank sampling instruments that were then carried all the way through the amplicon sequencing process).

Define a generic column with TRUE for all negative controls, and FALSE for real (positive) samples.

```{r}
df$NegativeControl <- grepl("Negative", df$SampleType)
table(df$SampleType, df$NegativeControl)
```

Inspect the DNA concentrations.

```{r}
class(df$DNA_Conc)
table(df$DNA_Conc, useNA="ifany")
```

`$DNA_Conc` is mostly ND (i.e. below limit of detection), and is currently in character format since its a mix of numbers and ND strings.

Define a new numeric `$conc.dna` column with the NDs replaced by zeroes.

```{r}
df$conc.dna <- df$DNA_Conc
df$conc.dna[df$conc.dna == "ND"] <- 0
df$conc.dna <- as.numeric(df$conc.dna)
summary(df$conc.dna)
```

```{r}
table(df$SampleType, df$conc.dna>0, useNA="ifany")
```

DNA above the limit of detection occurs at a higher rate in the Positive (real) samples, which is good, and the one NA concentration corresponds with the negative library sample (as it should).

```{r}
class(df$Library_Conc)
summary(df$Library_Conc)
```

Quite different. Every sample has a detectable concentration after libraries are normalized to try to produce equimolar amounts of DNA.

```{r}
ggplot(data=df, aes(x=conc.dna, y=Library_Conc, color=SampleType)) + 
  geom_point() + 
#  scale_x_log10() + scale_y_log10() +
  xlab("DNA Concentration (sample)") + ylab("DNA Concentration (library)")
```

In this sample set, where DNA concentrations of so many samples are initially below the limit of detection, the association between DNA concentration before and after library construction is fairly weak.

In sum, because such a large fraction of the pre-normalization concentrations are ND, and that is the more informative value for using the "frequency" method to detect contaminants, there is limited utility for frequency-based contaminant identification here.

## Curate samples

Check the distribution of library sizes (read depths) for samples:

```{r}
if(!identical(rownames(df), rownames(st))) stop("Mismatched samples in df and st.")
df$LibrarySize <- rowSums(st[rownames(df),])
ggplot(data=df, aes(x=rank(LibrarySize), y=LibrarySize, color=SampleType)) + geom_point()
```

```{r}
sort(df$LibrarySize)[1:10]
```

Three very low depth samples (<200 reads). The rest all have >5000 reads.

Remove the low-depth samples.

```{r}
if(!identical(rownames(df), rownames(st))) stop("Mismatched samples in df and st.")
sam <- rownames(df)[df$LibrarySize > 500]
df <- df[sam,]
st <- st[sam,]
```

```{r}
table(df$SampleType, df$NegativeControl)
```

## Curate ASVs

Use the distribution of ASV prevalences and abundances to guide feature-wise pre-filtering.

Inspect the distribution of ASV prevalences.
```{r}
table(colSums(st>0))
```

A very large number of ASVs that appear in only one or fewer samples.

Create a data.frame with rows for each ASV, and columns for prevalence and abundance across all samples, and only in positive samples.

```{r}
df.asv <- data.frame(Sequence=colnames(st), 
                     Prevalence=colSums(st>0),
                     Prevalence.Pos=colSums(st[!df$NegativeControl,]>0),
                     Abundance=colSums(st),
                     Abundance.Pos=colSums(st[!df$NegativeControl,]))
```

Inspect joint distribution of prevalence and abundance.

```{r}
ggplot(data=df.asv, aes(x=Prevalence, y=Abundance)) + 
  geom_point() + 
  scale_y_log10()
```

Inspect joint distribution of prevalence and abundance (again), but this time in only the positive samples.

```{r}
ggplot(data=df.asv, aes(x=Prevalence.Pos, y=Abundance.Pos)) + geom_point() + scale_y_log10()
```

There are a small number of higher abundance but low prevalence ASVs, even when considering just the positive samples.

*Choice of filtering threshold*: Presence in at least *5* positive samples OR at least *1000* total reads in positive samples (to be permissive to potential taxa that might rarely, but truly, occur in real samples). That's a 5/57 ~ 9% prevalence threshold.

```{r}
st.filt <- st[,df.asv$Prevalence.Pos >= 5 | df.asv$Abundance.Pos >= 1000]
ncol(st.filt); ncol(st); ncol(st.filt)/ncol(st)
```

```{r}
sum(st.filt)/sum(st)
```

This filtering reduced the number of ASVs from ~11k to ~750, but kept >90% of total reads.

Assign taxonomies using Silva v138.1 for these ASVs.

```{r}
### Uncomment the next two lines to regenerate the taxonomy table anew
### tax.filt <- assignTaxonomy(st.filt, "~/tax/silva_nr99_v138.1_train_set.fa.gz", multi=TRUE)
### saveRDS(tax.filt, "../RDS/tax_filt.rds")
tax.filt <- readRDS("../RDS/tax_filt.rds")
if(!identical(getSequences(tax.filt), getSequences(st.filt))) stop("Table mismatch: tax.filt, st.filt")
```


```{r}
table(tax.filt[,"Phylum"], useNA="ifany")
```

Phylum results look roughly in accordance with what would be expected from instestinal bacteria.

```{r}
tapply(colSums(st.filt), tax.filt[,"Phylum"], sum)/sum(st.filt)
```

Also at the read level: Bacteroidota and Firmicutes dominate the total abunbdance, with a good number of Actinobacteriota and Proteos as well (~10\% of each). All others at `1\%` or much less.

Save the curated data in R format.

```{r}
if(!identical(getSequences(tax.filt), getSequences(st.filt))) stop("Table mismatch: tax.filt, st.filt")
if(!identical(rownames(df), rownames(st))) stop("Mismatched samples in df and st.")
if(!identical(rownames(df), rownames(st.filt))) stop("Mismatched samples in df and st.")
save(st, st.filt, tax.filt, df, file="../RDS/bile.rda")
```

```{r}
sessionInfo()
```








